### DISCUSSION TITLE
Adding filters to a subquery, while having the support to reference the columns of the subquery

### DISCUSSION THREAD

**Discussion by sebtrack** [sqlalchemy/sqlalchemy#7960] [Usage Questions] [answered]
Hi there!

I'm working on a composable query architecture. We have many queries that repeat, run on their own or in joins. We had them all defined repeatedly, even moved away from SQLAlchemy to raw sql, because it was hard to debug.

I'm working on a new architecture to have them defined statically but the filters should be applicable dynamically. I have a second way doing it all dynamically (it's a bit more verbose but works fine as well, what happens statically in the example can happen in the constructor), but the problem, the discrepancy between the order of column definition and filter definition, stays similar.

```python
class CountAQuery(ModelQueryWrapper):
    model = AModel
    selected_columns = [model.col1, model.col2]
    selected_aggregations = [func.count(func.distinct(model.col1)).label("unique_count")]

class CountBQuery(ModelQueryWrapper):
    model = BModel
    selected_columns = [model.col1, model.col2]
    selected_aggregations = [func.count(model.col1).label("all_count")]

### until here everything works perfectly, because the models have clear attr/column definitions

class JoinABQuery(QueryWrapper):
    A = CountAQuery("A")
    B = CountBQuery("B")
    
    frequency = func.coalesce(cast(A.q.c.unique_count, sqla.Float)/ func.nullif(B.q.c.all_count, 0), 0).label("frequency")
    
    selected_columns = []
    selected_aggregations = [frequency]

    def _q(self, q):
        q1 = self.A.q
        q1 = q1.filters(self.A.model.token == "123")
        q2 = self.B.q
        q2 = q2.filters(self.B.model.token == "123")


        join_q = (
            q2.join(q1,
               (q2.c.analytics_token == q1.c.analytics_token))
        )

        q = q(*self.get_selected()).select_from(join_q)

        if self.subquery_name is not None:
            q = q.subquery(self.subquery_name)
        return q

```

When I have a query that I'd like to use in a join, I have to 'finalize' the query with `.subquery(...)`. Once I do it, I can't add filters, as `'Alias' object has no attribute 'filters'`. Conceptually I thought it shouldn't make any difference adding filters afterwards, as it does not change the columns.

When I don't "finalize" with subquery I can return a BaseQuery, which is awesome as all my subclasses can do whatever they want with the query. But this makes defining the aggregations and columns hard (these are clear at non-runtime), as it always has to happen after I've added the filters (which come in in runtime).

Is there any chance, to achieve what I want? Maybe a way to extract the `Selectable` from the `Alias`, add filters to it and put them back into a `Alias`? Or using strings to reference the columns? (I know that you keep references to the object as I was running into a nasty bug, which was repeating select queries in my joins, when I was constructing the query on each call.) I saw an "adapt_on_names" in `AliasedClass` but it's not exposed for subquery calls, does it solve my issue and can I work around the `.subquery` call?

We run v1.3.22 and can't migrate for now.

Any help very much appreciated!

**Comment by zzzeek** [comment, marked as answer]
Hi there -

this is hard for me to follow because it's missing lots of detail:

* What is `filters()`  ?  that's not any method in SQLAlchemy, did you mean `filter()` ?
* What kind of object is `A.q` and `B.q` ?  these seem like subqueries, because you are calling `.c` upon them, but then you are also calling `.filters()` (which I am guessing is filter()?) on them, but that method is not on `Alias` in SQLAlchemy 1.3, so I dont understand what you're doing (unless that's the problem?)
* What kind of object is passed to `def _q()`, is that Query?

my attempt at an answer: since you are building up the Query as you go, and you want to keep it buildable, then obviously, don't call subquery() until you are done changing it.   add the "aggregations" using a translation function, change "selected_aggregations" to a callable that is given the subquery that you want to aggregate and does the aggregations at that point.    It's hard for me to see any particular nuance that might be making this more difficult with the info given.

**Reply by sebtrack** [reply]
1) Oh yes I meant `filter`, sorry went wrong when I tried to make the example.
2) `q` is calling `_q`. As in the example, it was meant to just pass the `BaseQuery` but due to the "subquery" issue it was also calling subquery if a name was provided. So yes this is part of my problem.
3) the q passed to _q is either a db.session.query or a Model.query.

Like this?
```
frequency = lambda A,B: func.coalesce(cast(A.q.c.unique_count, sqla.Float)/ func.nullif(B.q.c.all_count, 0), 0).label("frequency")
    
selected_columns = []
selected_aggregations = [frequency]
```

This is a good idea, it delays the evaluation to a point where I can add the filters and still have a static definition. Initially I was checking if all aggregations had labels in the Metaclass, this isn't possible anymore.

Thanks a lot. If you still have an even better idea. I'd love to hear about it. (As SqlAlchemy uses a lot of the Metaclass magic).

**Reply by zzzeek** [reply]
yes, but better still use a real def and make it either `@classmethod` or `@staticmethod`.     lambdas as class variables doesn't go very far in modern Python tooling.

**Reply by sebtrack** [reply]
Yeah, I was thinking to wrap the QueryWrappers (the subqueries) into it's own managing class and then have the columns be wrapped as well so that the subqueries (QueryWrappers) always get matched to its corresponding columns.

To sketch it out:

```python
class JoinABQuery(QueryWrapper):
    subqueries = SubQueries(A=CountAQuery, B=CountBQuery)
    frequency = Column(lambda A,B, **kwargs: func.coalesce(cast(A.q.c.unique_count, sqla.Float)/ func.nullif(B.q.c.all_count, 0), 0).label("frequency"))
)
```

Then let `SubQueries` provide `kwargs` to all column objects, and each definition can use the names the developer initially gave it.
